### PR TITLE
WIP:  apache for loadbalancing

### PR DESCRIPTION
**Pull request by paulczar**
This is a quick POC / WIP to demo using apache instead of haproxy for loadbalancing our APIs.

I think this gives us more flexibiltiy as we can enable in the APIs role itself ( in this case in keystone ) rather than a big monolithic haproxy.conf.      We also already have apache running for horizon, so it could reduce number of services.    apache also has better logging options that haproxy which can only log to syslog.

### CODE REVIEW

File: library/apache2_site [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    if re.match(r'.*' + name + r' already disabled.*', stdout, re.S):
+        site.exit_json(changed = False, result = "Success")
+    elif result != 0:
+        site.fail_json(msg="Failed to disable site %s: %s" % (name, stdout))
+    else:
+        site.exit_json(changed = True, result = "Disabled")
+
+def _enable_site(site):
+    name = site.params['name']
+    a2dissite_binary = site.get_bin_path("a2ensite")
```
Comment (j2sol) [inline library/apache2_site]: Looks like copypasta. a2dissite_binary in the enable_site function.

File: library/apache2_site [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+# disables the Apache2 site "default"
+- apache2_site: state=disabled name=default
+'''
+
+import re
+
+def _disable_site(site):
+    name = site.params['name']
+    a2dissite_binary = site.get_bin_path("a2dissite")
```
Comment (j2sol) [inline library/apache2_site]: What happens when a2dissite (and a2ensite) isn't available? Can you have the module handle that a bit gracefully?

File: roles/apache/handlers/main.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+- name: restart apache
+  service: name=apache2 state=restarted
+
+- name: reload apache
+  service: name=apache2 state=reloaded
```
Comment (j2sol) [inline roles/apache/handlers/main.yml]: What happens if we have a restart and a reload triggered? Does apache handle this well, sending one right after another? This is what got us in trouble with haproxy.

File: roles/apache/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    - ssl
+    - proxy
+    - proxy_ajp
+    - proxy_http
+    - rewrite
+    - deflate
+    - headers
+    - proxy_balancer
+    - proxy_connect
+    - proxy_html
```
Comment (j2sol) [inline roles/apache/tasks/main.yml]: nit pick, can you alpha sort the module list?  :laughing:

File: roles/apache/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  notify:
+    - reload apache
+
+- name: disable default apache site
+  apache2_site: state=disabled name=000-default
+  notify:
+    - reload apache
+
+- name: check if custom ports.conf
+  exit_status: grep '# This file is managed by Ansible' /etc/apache2/ports.conf
```
Comment (j2sol) [inline roles/apache/tasks/main.yml]: can we not use exit_status? It's a dep of antsy, and ruby, and it's custom, and we have to carry it forward, and we can achieve the same thing with ansible built ins such as failed_when

File: roles/apache/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  run_once: true
+  register: custom_apache_ports_file
+
+- name: create custom ports.conf file
+  template: src=etc/apache2/ports.conf
+            dest=/etc/apache2/ports.conf
+            owner=root
+            group=root
+            mode=0644
+  when: custom_apache_ports_file.status != 0
```
Comment (j2sol) [inline roles/apache/tasks/main.yml]: Do we expect things external to ansible to be manipulating the ports.conf file? Is that why we can't just template the file and call it good?

Comment (j2sol) [inline roles/apache/tasks/main.yml]: also, does anything need to be reloaded/restarted if the ports config changes?

File: roles/horizon/tasks/main.yml [hunk 7]
```diff
@@ -1,121 +1,92 @@
...
-
 - name: lesscpy must be in apache PATH
   pip: name=lesscpy version=0.9j
 
 - name: get horizon source repo
-  git: repo={{ openstack.git_mirror }}/horizon.git
-       dest=/opt/stack/horizon
-       version={{ horizon.rev }}
-       update={{ openstack.git_update }}
+  git: |
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: Shouldn't need the line continuation here. Just put the first key=value here and indent everything to that level below it.

Comment (j2sol) [inline roles/horizon/tasks/main.yml]: in fact, just how it was, why was this changed?

File: roles/horizon/tasks/main.yml [hunk 8]
```diff
@@ -1,121 +1,92 @@
...
+  lineinfile: dest=/etc/apache2/ports.conf line="Listen 8080"
   notify:
-    - restart apache
+    - reload apache
 
-- name: openstack dashboard config (12.04)
-  template: src=etc/apache2/sites-available/openstack_dashboard.conf
-            dest=/etc/apache2/sites-available/openstack_dashboard
+- name: create dashboard virtualhost on precise
+  template: |
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: why a line continuation here?

File: roles/horizon/tasks/main.yml [hunk 9]
```diff
@@ -1,121 +1,92 @@
...
   when: ansible_distribution_version == "12.04"
   notify:
-    - restart apache
+    - reload apache
 
-- name: openstack dashboard config
-  template: src=etc/apache2/sites-available/openstack_dashboard.conf
-            dest=/etc/apache2/sites-available/openstack_dashboard.conf
+- name: create dashboard virtualhost on other
+  template: |
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: why the continuation?

File: roles/horizon/tasks/main.yml [hunk 10]
```diff
@@ -1,121 +1,92 @@
...
-- name: static asset directories
-  file: dest={{ item }}
-        state=directory
-        owner=www-data
-        group=www-data
-        mode=0755
+    - reload apache
+
+- name: create static asset dirs
+  file: |
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: why so continuation?

File: roles/horizon/tasks/main.yml [hunk 11]
```diff
@@ -1,121 +1,92 @@
...
   with_items:
     - /opt/stack/horizon/static
     - /opt/stack/horizon/static/dashboard
 
-- name: horizon local settings
-  template: src=opt/stack/horizon/openstack_dashboard/local/local_settings.py
-            dest=/opt/stack/horizon/openstack_dashboard/local/local_settings.py
-            mode=0644
+- name: dashboard settings
+  template: |
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: ?

File: roles/horizon/tasks/main.yml [hunk 12]
```diff
@@ -1,121 +1,92 @@
...
+    mode=0644
   notify:
-    - restart apache
+    - reload apache
 
 - name: custom horizon logo
-  get_url: url={{ horizon.logo_url }}
-           dest=/opt/stack/horizon/openstack_dashboard/static/dashboard/img/logo.png
-           mode=0644 force=yes
+  get_url: url={{ horizon.logo_url }} dest=/opt/stack/horizon/openstack_dashboard/static/dashboard/img/logo.png mode=0644 force=yes
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: Why did this go single line?

File: roles/horizon/tasks/main.yml [hunk 13]
```diff
@@ -1,121 +1,92 @@
...
-  get_url: url={{ horizon.logo_url }}
-           dest=/opt/stack/horizon/openstack_dashboard/static/dashboard/img/logo-splash.png
-           mode=0644 force=yes
+  get_url: url={{ horizon.logo_url }} dest=/opt/stack/horizon/openstack_dashboard/static/dashboard/img/logo-splash.png mode=0644 force=yes
 
 - name: custom horizon favicon
-  get_url: url={{ horizon.favicon_url }}
-           dest=/opt/stack/horizon/openstack_dashboard/static/dashboard/img/favicon.ico
-           force=yes
+  get_url: url={{ horizon.favicon_url }} dest=/opt/stack/horizon/openstack_dashboard/static/dashboard/img/favicon.ico force=yes
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: Same for the other two get_urls above.

File: roles/horizon/tasks/main.yml [hunk 14]
```diff
@@ -1,121 +1,92 @@
...
 
 - name: put images and fonts where apache can find them
-  file: src=/opt/stack/horizon/openstack_dashboard/static/dashboard/{{ item }}
-        dest=/opt/stack/horizon/static/dashboard/{{ item }}
-        state=link
-        owner=www-data
-        group=www-data
+  file: |
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: ?

File: roles/horizon/tasks/main.yml [hunk 15]
```diff
@@ -1,121 +1,92 @@
...
+    group=www-data
   with_items:
     - img
     - fonts
 
-# flush before ensuring apache running, saves immediate restart
-- meta: flush_handlers
-
-- name: ensure apache started
+- name: ensure apache is running
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: Now you have a start followed by an immediate reload. Will apache be okay with that?

File: roles/horizon/tasks/main.yml [hunk 16]
```diff
@@ -1,121 +1,92 @@
...
 
-- name: Permit HTTP and HTTPS
-  ufw: rule=allow to_port={{ item }} proto=tcp
-  with_items:
-  - 80
-  - 443
+- name: Permit HTTP (redirect to HTTPS for Horizon)
+  command: ufw allow 80/tcp
 
-- include: monitoring.yml tags=monitoring,common
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: What happened to the monitoring tasks?

File: roles/horizon/tasks/main.yml [hunk 17]
```diff
@@ -1,121 +1,92 @@
...
+- name: Permit HTTPS (Horizon)
+  command: ufw allow 443/tcp
```
Comment (j2sol) [inline roles/horizon/tasks/main.yml]: Why did this go from a single task with list of ports to two tasks?

File: roles/keystone/templates/etc/apache2/sites-available/keystone-admin.conf [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    ProxyPass / balancer://keystone
+
+    Include /etc/apache2/ssl.conf
+    SSLCertificateFile /etc/apache2/ssl.crt
+    SSLCertificateKeyFile /etc/apache2/ssl.key
+</VirtualHost>
+
+<Proxy balancer://keystone>
+{% for host in groups['controller'] -%}
+  BalancerMember http://{{ hostvars[host][primary_interface]['ipv4']['address'] }}:35357/
```
Comment (j2sol) [inline roles/keystone/templates/etc/apache2/sites-available/keystone-admin.conf]: Shouldn't this be 35358? or do we have keystone locally listening on 35357 (the default), and only front end 35358 in apache?

File: roles/keystone/templates/etc/apache2/sites-available/keystone-admin.conf [hunk 19]
```diff
@@ -0,0 +1,17 @@
+<VirtualHost *:35358>
+    LogLevel  info
+    ErrorLog  /var/log/apache2/keystone-error.log
+    CustomLog /var/log/apache2/keystone-access.log combined
+    Options +FollowSymLinks
```
Comment (j2sol) [inline roles/keystone/templates/etc/apache2/sites-available/keystone-admin.conf]: Does keystone need to follow symlinks?

File: roles/apache/tasks/main.yml (hunk 5 above)
Comment (j2sol) [inline roles/apache/tasks/main.yml]: In fact, exit_status is being removed in #670

**Comment by paulczar** [comment]
Hey folks,  this was a really quick POC thrown together to demonstrate the concept,  don't nitpick the node just yet,  just discuss the idea of using apache to do the http load balancing over haproxy.

**Comment by j2sol** [comment]
I have a question, and this is related to either haproxy or apache as the LB, but if I wanted to do a rolling restart of services, I generally need to disable one of the end points at the load balancer and allow it to "drain" any active connections. Otherwise unless the service check is EXTREMELY aggressive, we run the risk of sending connections to a node that has a downed service. With LBaaS there is typically an API call one can make to enact this. Does anything like that exist for apache (or haproxy) as our LB, and if not, is there a suggested strategy to ensure the LB doesn't route any connections to a node where we're restarting services?

**Comment by j2sol** [comment]
Looks like you can do run-time adjustments to the pool by way of balance-manager, but that has to be enabled.  http://httpd.apache.org/docs/2.2/mod/mod_proxy_balancer.html speaks of it. Is this something we'd want to expose at least locally so that we can do orchestrated service restarts?

**Comment by paulczar** [comment]
My personal opinion on this, is that these are low traffic API endpoints, and the occasional mis-route during maintenance is unlikely to cause any real issue apart from the user having to refire the API call which they should get a decent error message about.   so it may not be worth a /lot/ of engineering time to ensure absolute zero downtime.

**Comment by j2sol** [comment]
That's a fair point. I was mostly curious if using the apache load balancer would even allow for it.

**Pull request closed without merging**